Healpix (#605)
cf-convention/cf-conventions: appf.adoc

The coordinate system is defined by discrete cells on a sphere, arranged on a HEALPix grid.
Each cell is represented by its cell index, which, together with the `indexing_scheme` and possible `refinement_level`, determines center coordinates (latitude and longitude) as well as boundaries.
See <<GHB05>> for the calculation of center coordinates and cell boundaries coordinates.

__Notes:__::

Cells are indexed using either the nested, ring, nuniq or zuniq indexing scheme, the indices are stored in the index coordinate, which can be identified through the standard name `healpix_index`, or the `grid_mapping` attribute.

For "nested" and "ring" scheme, the grid resolution is controlled by the refinement level parameter **`refinement_level`**, for "nuniq" and "zuniq", the resolution of a cell is implied by the index.

An index according to the `nuniq` indexing scheme can be computed from the `refinement_level` and a `nested` index as follows <<RH15>>:

`index_nuniq = index_nested + 4 ^ (1 + refinement_level)`

> TODO: add computation for `zuniq`. This seems to depend on `DEPTH_MAX`, which seems to depend on the number of available bits. How do we (formally) define DEPTH_MAX?

Reviewer: > Do we have another metadata field?

I would say we do not need another field. I do not see any reason to encode `zuniq` in a weird custom integer type. Only if somebody thinks that HEALPix is still useful below cm resolution and therefore needs more than 64-bits which I doubt.

> Do we state that this is always the number of bits of the integer data type used?

I would say yes and connect the number of bits with the integer data type used within the netCDF file and recommend 64-bit integers for non-expert users. Even if we only store cells at a level below 13 which would fit into a 32-bit integer, using the shuffle filter with compression would shrink these zeros away, so there is not even an argument on saving space, but I do not want to take away the possibility to use only 32-bit for expert users who want to load and directly stream data in the most optimized way.

> How does this play with e.g. Python data types (where numbers are of arbitrary size)?

I would delegate the responsibility of keeping track of data types to the applications. Most users in Python will likely use libraries like NumPy to load such data, so they will likely be in a typed ecosystem anyway.

> How does this play with e.g. tie-point coordinates, where technically, the cell index values are interpolated, and the data type of the interpolated variable is (at least in theory) dependent on the client-side implementation.

The interpolation has the same limitation as for the nested index that if it is floating point based, it will be limited by the number of significand bits. An IEEE 754 double-precision float has 52 significand bits. Therefore, it can only interpolate down to `Lmax = (52-4-1)//2 = 23`. In case of `zuniq`, the trailing zeros are picked up by the exponent.
However, this is likely more a problem with `nested` indexing, because `zuniq` is more often used for MOCs which combine consecutive indices to lower-level cells already.

I think, we may somehow put a warning sign about the tie-point interpolation somewhere. In the worst case, delta compression can still help avoid wasting storage for the index.

> TODO: add computation for `zuniq`. This seems to depend on `DEPTH_MAX`, which seems to depend on the number of available bits. How do we (formally) define DEPTH_MAX?

Reviewer: We may also go with the s2geometry approach and define the `zuniq` index with 64-bit integers. An expert user is not forced to follow cf-conventions... What do you think?

+
The coordinate system is defined by discrete cells on a reference ellipsoid (or sphere), arranged on a HEALPix grid.
Each cell is represented by its cell index, which, together with the `indexing_scheme` and possible `refinement_level`, determines center coordinates (latitude and longitude) as well as boundaries.
See <<GHB05>> for the calculation of center coordinates and cell boundary coordinates.

__Notes:__::

Cells are indexed using the scheme identified by the **`indexing_scheme`** parameter.
+
For both the **`nested`** and **`ring`** schemes, the grid resolution is controlled by the refinement level parameter **`refinement_level`** <<GHB05>>.
The total number of cells on the surface of the sphere (or reference ellipsoid) is `12 * (4 ^ refinement_level)`.
The lowest allowed refinement level is `0`, giving 12 cells globally.

> The algorithm that assigns cell locations to **`nested`** or **`ring`** indices at a given refinement level is defined in <<GHB05>>. In this algorithm, a cell with index `0`, at any refinment level, always has its southern vertex on the meridian at 45 degrees east.

Author: If this helps (regarding `a` and `b`), there's already a [definition for `semi_major_axis` and `semi_minor_axis` in CF-conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#appendix-grid-mappings), and as far as I get the current state, these parameters would have to be specified anyways, if users want to communicate, that any grid mapping is using an ellipsoid.

So if we actually specify that we are always using authalic latitudes for HEALPix computations, I guess we can take these two parameters for granted (always, as it doesn't matter for a sphere).

> The algorithm that assigns cell locations to **`nested`** or **`ring`** indices at a given refinement level is defined in <<GHB05>>. In this algorithm, a cell with index `0`, at any refinment level, always has its southern vertex on the meridian at 45 degrees east.

Reviewer 2: ```suggestion
The algorithm that assigns cell locations on a sphere to **`nested`** or **`ring`** indices at a given refinement level is defined in <<GHB05>>.
In this algorithm, a cell with index `0`, at any refinement level, always has its southernmost vertex on the meridian at 45 degrees east.
Cell locations on the surface of an ellipsoid are defined to give equal-area HEALPix cells when using authalic latitudes (which map locations on an ellipsoid to those on the sphere with the same surface area).
For instance, on the WGS84 ellipsoid this results in the center of the HEALPix cell at refinement level `0` with index `0` being at geodetic latitude 41.938 degrees north, instead of at 41.810 degrees north in the case of a sphere, whilst the longitude is at 45 degrees east in both cases.
```

The use of `projection_x_coordinate` and `projection_y_coordinate` was deprecated in version 1.9 of the CF Conventions.
The initial definition of this projection used these standard names to identify the projection coordinates even though their canonical units (meters) do not match those required for this projection (radians).

[[healpix]]
=== HEALPix

----
grid_mapping_name = healpix
----

__Map parameters:__::
* **`indexing_scheme`** - The HEALPix indexing scheme: One of **`nested`**, **`ring`**, **`nuniq`**, or **`zuniq`**
* **`refinement_level`** - The HEALPix refinement level: Mandatory for indexing scheme **`nested`** or **`ring`**, but must be omitted for indexing scheme **`nuniq`** or **`zuniq`**

__Map coordinates:__::
The coordinate system is defined by discrete cells on a reference ellipsoid (or sphere), arranged on a HEALPix grid.
Each cell is represented by its cell index, which, together with the `indexing_scheme` and possible `refinement_level`, determines center coordinates (latitude and longitude) as well as boundaries.
See <<GHB05>> for the calculation of center coordinates and cell boundary coordinates.
+

> The integer-valued HEALPix index coordinates are identified by the **`standard_name`** attribute value **`healpix_index`**.

Author: They are either non-negative, or it doesn't matter (because the number is interpreted as bit-pattern and the sign-bit isn't used). But in any case, the non-negative version seems to be the more accepted and elegant version. Thus I think we could make it a conformance requirement.

> The HEALPix index coordinates may be stored in a coordinate, auxiliary, or scalar coordinate variable. Indexing schemes are optimized to allow applications efficient data access when the HEALPix indices are ordered strictly monotonically. Therefore it is recommended to store strictly monotonic HEALPix indices in a coordinate variable, which allows an application to assess their monotonicity without expensive checking.

Reviewer 3: We usually take a new line for each sentence in `adoc`

This is necessary to retain the equal area property.
For instance, on the WGS84 ellipsoid this results in the center of the HEALPix cell at refinement level `0` with index `0` being at geodetic latitude 41.938 degrees north, instead of at 41.810 degrees north in the case of a sphere, whilst the longitude is at 45 degrees east in both cases.
+
The **`ring`** scheme sorts pixels along iso-latitude rings from north to south; and **`nested`** indexing sorts pixels along the Z-order curve within each of the 12 global cells at refinement level `0`, resulting in spatial proximity for consecutive pixels.
+
The **`nuniq`** and **`zuniq`** indexing schemes allow pixels from multiple refinement levels to be combined: **`nuniq`** effectively concatenates **`nested`** indices from lower to the higher refinement levels, resulting in spatial proximity for consecutive pixels within a refinement level but not across different refinement levels; and **`zuniq`** indexing sorts pixels of any refinement level along the Z-order curve, resulting in spatial proximity for consecutive pixels, regardless of their refinement level.
+

> An index according to the **`nuniq`** indexing scheme can be computed from a refinement level and a **`nested`** index with the formula `index_nuniq = index_nested + 4 ^ (1 + refinement_level)` <<RH15>>.

Author: I tried to clarify the sentence.

If tie-point coordinates are used, take special care of the required precision.

+
The coordinate system is defined by discrete cells on a sphere, arranged on a HEALPix grid.
Each cell is represented by its cell index, which, together with the `indexing_scheme` and possible `refinement_level`, determines center coordinates (latitude and longitude) as well as boundaries.
See <<GHB05>> for the calculation of center coordinates and cell boundary coordinates.

__Notes:__::

Cells are indexed using either the nested, ring, nuniq or zuniq indexing scheme.

+
For "nested" and "ring" scheme, the grid resolution is controlled by the refinement level parameter **`refinement_level`** <<GHB05>>.
0 is the coarsest, the total number of cells on the sphere is `12 * 4 ^ refinement_level`.
For the "nuniq" and "zuniq" schemes, the resolution of a cell is implied by the index, and `refinement_level` must be omitted.

+
An index according to the `nuniq` indexing scheme can be computed from the `refinement_level` and a `nested` index as follows <<RH15>>:

+
`index_nuniq = index_nested + 4 ^ (1 + refinement_level)`

+

> The `zuniq` indexing scheme depends on the number of bits used by the integer type.

Reviewer 2: I've had a go at incorporating your text, @carloshorn, which rearranged things a bit. The content's basically the same as before, with Carlos's text, and I've added a note about cell `0` being at 45 degrees east.

The text is here rendered in markdown, with the raw asciidoc below

----

Cells are indexed using either the **`nested`**, **`ring`**, **`nuniq`** or **`zuniq`** indexing scheme.

The **`ring`** scheme sorts pixels along iso-latitude rings from north to south; and **`nested`** indexing sorts pixels along the Z-order curve, resulting in spatial proximity for consecutively pixels.

For both the **`nested`** and **`ring`** schemes, the grid resolution is controlled by the refinement level parameter **`refinement_level`** <<GHB05>>.
The total number of cells on the sphere is `12 * (4 ^ refinement_level)`.
The lowest allowed refinement level is `0`, giving 12 cells globally.
The algorithm that assigns cell locations to **`nested`** or **`ring`** indices at a given refinement level is defined in <<GHB05>>. In this algorithm, a cell with index `0`, at any refinment level, always has its southern vertex on the meridian at 45 degrees east.

The **`nuniq`** and **`zuniq`** indexing schemes allow pixels from multiple refinement levels to be combined: **`nuniq`** effectively concatenates **`nested`** indices from lower to the higher refinement levels, resulting in spatial proximity for consecutive pixels within a refinement level but not across different refinement levels; and **`zuniq`** indexing sorts pixels of any refinement level along the Z-order curve, resulting in spatial proximity for consecutive pixels, regardless of their refinement level.

An index according to the **`nuniq`** indexing scheme can be computed from a refinement level and a **`nested`** index with the formula `index_nuniq = index_nested + 4 ^ (1 + refinement_level)` <<RH15>>.

An index according to the **`zuniq`** indexing scheme can be computed from a refinement level and a **`nested`** index with the formula `index_zuniq = (2 * index_nested + 1) * 4 ^ (29 - refinement_level)` <<K24>>.

The  **`refinement_level`** parameter must be omitted for the **`nuniq`** and **`zuniq`** schemes, for which the resolution of a cell is implied by the index alone.

A refinement level, either given by the  **`refinement_level`** parameter or implied by **`nuniq`** and **`zuniq`** indices, must be no greater than `29`, which is the highest refinement level for which all indices of any indexing scheme can be represented by 64-bit integers.

----

```markdown
Cells are indexed using either the **`nested`**, **`ring`**, **`nuniq`** or **`zuniq`** indexing scheme.

The **`ring`** scheme sorts pixels along iso-latitude rings from north to south; and **`nested`** indexing sorts pixels along the Z-order curve, resulting in spatial proximity for consecutively pixels.

For both the **`nested`** and **`ring`** schemes, the grid resolution is controlled by the refinement level parameter **`refinement_level`** <<GHB05>>.
The total number of cells on the sphere is `12 * (4 ^ refinement_level)`.
The lowest allowed refinement level is `0`, giving 12 cells globally.
The algorithm that assigns cell locations to **`nested`** or **`ring`** indices at a given refinement level is defined in <<GHB05>>. In this algorithm, a cell with index `0`, at any refinment level, always has its southern vertex on the meridian at 45 degrees east.

The **`nuniq`** and **`zuniq`** indexing schemes allow pixels from multiple refinement levels to be combined: **`nuniq`** effectively concatenates **`nested`** indices from lower to the higher refinement levels, resulting in spatial proximity for consecutive pixels within a refinement level but not across different refinement levels; and **`zuniq`** indexing sorts pixels of any refinement level along the Z-order curve, resulting in spatial proximity for consecutive pixels, regardless of their refinement level.

An index according to the **`nuniq`** indexing scheme can be computed from a refinement level and a **`nested`** index with the formula `index_nuniq = index_nested + 4 ^ (1 + refinement_level)` <<RH15>>.

An index according to the **`zuniq`** indexing scheme can be computed from a refinement level and a **`nested`** index with the formula `index_zuniq = (2 * index_nested + 1) * 4 ^ (29 - refinement_level)` <<K24>>.

The  **`refinement_level`** parameter must be omitted for the **`nuniq`** and **`zuniq`** schemes, for which the resolution of a cell is implied by the index alone.

A refinement level, either given by the  **`refinement_level`** parameter or implied by **`nuniq`** and **`zuniq`** indices, must be no greater than `29`, which is the highest refinement level for which all indices of any indexing scheme can be represented by 64-bit integers.
```

The use of `projection_x_coordinate` and `projection_y_coordinate` was deprecated in version 1.9 of the CF Conventions.
The initial definition of this projection used these standard names to identify the projection coordinates even though their canonical units (meters) do not match those required for this projection (radians).

[[healpix]]
=== HEALPix

----
grid_mapping_name = healpix
----

__Map parameters:__::
* **`indexing_scheme`**
* **`refinement_level`** - Mandatory for `indexing_scheme` "nested" or "ring", must be omitted for "nuniq" or "zuniq".

__Map coordinates:__::
The map uses a discrete 1-dimensional "index" coordinate, which uses the `standard_name` `"healpix_index"`, and can additionally be identified through the `grid_mapping` attribute.
The coordinate can be any kind of coordinate, apart from coordinate variables or auxiliary coordinates, especially a tie-point coordinate can be used to encode contiguous blocks in index space in a compact way.

> If tie-point coordinates are used, take special care of the required precision.

Author: I agree. While I still think, tie point coordinates are super helpful, we probably shouldn't advertise them actively before solving the little details. And removing the comment doesn't forbid to use it :-)

+
The coordinate system is defined by discrete cells on a sphere, arranged on a HEALPix grid.
Each cell is represented by its cell index, which, together with the `indexing_scheme` and possible `refinement_level`, determines center coordinates (latitude and longitude) as well as boundaries.
See <<GHB05>> for the calculation of center coordinates and cell boundary coordinates.

__Notes:__::

> Cells are indexed using either the nested, ring, nuniq or zuniq indexing scheme, the indices are stored in the index coordinate, which can be identified through the standard name `healpix_index`, or the `grid_mapping` attribute.

Reviewer 2: ```suggestion
Cells are indexed using either the nested, ring, nuniq or zuniq indexing scheme.
```
No need to repeat the bit in "Map coordinates"

+

> For "nested" and "ring" scheme, the grid resolution is controlled by the refinement level parameter `refinement_level`.

Reviewer 2: ```suggestion
For "nested" and "ring" scheme, the grid resolution is controlled by the refinement level parameter **`refinement_level`** <<GHB05>>.
```

0 is the coarsest, the total number of cells on the sphere is `12 * 4 ^ refinement_level`.
For the "nuniq" and "zuniq" schemes, the resolution of a cell is implied by the index, and `refinement_level` must be omitted.

+
An index according to the `nuniq` indexing scheme can be computed from the `refinement_level` and a `nested` index as follows <<RH15>>:

+
`index_nuniq = index_nested + 4 ^ (1 + refinement_level)`

+
The `zuniq` indexing scheme depends on the number of bits used by the integer type.

> This specification assumes 64 bit integers for all `zuniq` indices.

Reviewer 2: ```suggestion
```
We already restrict refinement_level <= 29, so I don't see the advantage in talking about bits and precision, here.

An index according to the `zuniq` indexing scheme can be computed from the `refinement_level` and a `nested` index as follows <<K24>>:

+
`index_zuniq = (2 * index_nested + 1) * 4 ^ (29 - refinement_level)`

+

> The value `29` is the maximum representable `refinement_level` and is computed from the number of bits as `(n_bits - 4 - 1) // 2`.

Reviewer 2: ```suggestion
The value `29` is the maximum representable `refinement_level` (which is the highest level for which all indices can be represented by 64-bit integers).
```
What about `unsigned 64-bit`? Are we OK with that ... CF does not allow anything larger.

// The cleanest-looking, which sometimes works for asciidoctor-pdf:
     **`-90.0 \<= latitude_of_projection_origin \<= 90.0`**

| **`refinement_level`** | N

> | DGGS numerical order of a discrete global grid in the tessellation sequence. See "4.37. refinement level" in <<OGC_DGGS>>.

Reviewer: After the conversation about relating DGGRS -> DGGS -> DGGH, I am not sure about using an abbreviation in this text. Can we just call it the level within the cell hierarchy while keeping the reference for providing details in the DGG(R)S context?

| **`north_pole_grid_longitude`** | N
| Longitude (degrees) of the true north pole in the rotated grid.

| **`indexing_scheme`** | S

> | Identifier of the HEALPix cell indexing scheme, either `nested`, `ring`, `nuniq` or `zuniq`. See <<healpix>> for more detailed definitions.

Reviewer 2: ```suggestion
| The HEALPix cell indexing scheme, either **`nested`**, **`ring`**, **`nuniq`** or **`zuniq`**. See <<healpix>> for more detailed definitions.
```
